Optimize count leading ones if promoted type

v01dXYZ opened the pull request:
First suggestions to solve #96455.

tested with:

* scalar `X86`
* `AArch64`
*  `RISCV` (and `VP_`)

For `SelectionDAG` with `AMDGPU`, I tried to replace the custom legaliser for `CTLZ` (which is necessary to make convert it to `i32` instead of `i16` if no custom legaliser), but I didn't manage to get it work when there is a `ANYEXTEND` user. That's why the commit is reverted. `GlobalISel` works fine though.

Commit:
[CodeGen] Legalisation with promotion: optimise count leading ones

(CTLZ (XOR Op -1))
  -->
(CTLZ_ZERO_UNDEF (XOR (SHIFT (ANYEXTEND Op) ShiftAmount) -1))

The optimisation also applies for CTLZ_ZERO_UNDEF, VP_CTLZ, VP_CTLZ_ZERO_UNDEF.

Fixes https://github.com/llvm/llvm-project/issues/96455

arsenm commented on llvm/lib/CodeGen/CodeGenPrepare.cpp:
```diff
@@ -2298,8 +2298,29 @@ static bool despeculateCountZeros(IntrinsicInst *CountZeros,
+  if ((TLI->getTypeAction(CountZeros->getContext(), VTy) ==
+           TargetLowering::TypePromoteInteger ||
+       TLI->getOperationAction(ISD::CTLZ, VTy) == TargetLowering::Promote) &&
+      match(CountZeros->getOperand(0), m_Xor(m_Value(Op0), m_Value(Op1))) &&
+      ((C = dyn_cast<ConstantInt>(Op0)) || (C = dyn_cast<ConstantInt>(Op1))) &&
+      C->isMinusOne())
```
You can use m_Not instead of manually matching -1, which only works for scalars

arsenm commented on llvm/lib/CodeGen/GlobalISel/LegalizerHelper.cpp:
```diff
@@ -2356,6 +2356,35 @@ LegalizerHelper::widenScalarMulo(MachineInstr &MI, unsigned TypeIdx,
+  Register XorSrc;
+  Register CstReg;
+  if (!mi_match(MI.getOperand(1).getReg(), MRI,
+                m_GXor(m_Reg(XorSrc), m_Reg(CstReg))))
```
There should be an m_not helper like the IR if there isn't already

arsenm commented on llvm/lib/CodeGen/CodeGenPrepare.cpp:
```diff
@@ -2298,8 +2298,29 @@ static bool despeculateCountZeros(IntrinsicInst *CountZeros,
   Type *Ty = CountZeros->getType();
+  EVT VTy = TLI->getValueType(*DL, Ty);
+
+  // do not despeculate if we have (ctlz (xor op -1)) if the operand is
```
Should have IR test for this handling. The promotion check logic probably doesn't work correctly for vectors too

arsenm replied:
Capitalize comment. Also I guess given the vector check above, the vector case doesn't matter?

v01dXYZ replied:
I'll capitalise the comment but I don't understand why the check for TypeAction/OperationAction is not right in the case of a vector type (although here it's valid b/c previous check ensuring it's not a vector type).

v01dXYZ replied:
I made a mistake though since I should have used `InstructionOpcodeToISD`.

arsenm replied:
The vector case is skipped above, so I guess it doesn't matter.

The problem with checking the promotion logic like this, and in many other places, is there may be multiple type legalization steps before the operation is legal, so this isn't a 100% accurate way of determining the type this will ultimately use

dtcxzyw commented:
It is a bit weird that this patch contains both CGP and SDAG changes. Can we split it into multiple patches?

RKSimon commented on llvm/lib/CodeGen/SelectionDAG/LegalizeDAG.cpp:
```diff
@@ -5049,6 +5049,40 @@ static MVT getPromotedVectorElementType(const TargetLowering &TLI,
+  ConstantSDNode *Cst = dyn_cast<ConstantSDNode>(CstOp);
+
+  if (!Cst || !Cst->isAllOnes())
+    return false;
```
We can now use SDPatternMatch to do this for us - 
```
sd_match(Node->getOperand(0), m_Not(SrcOp))
```

Commit:
[CodeGenPrepare] Do not despeculate count leading/trailing ones if promotion

For count leading/trailing ones, ie (CTLZ/CTTZ (XOR Op -1)),
legalisation should be able to optimise this case when a promotion is
necessary.

Despeculation should not be applied in this case as it will separate
XOR and CTLZ/CTTZ in two different basic blocks. This is particularly
problematic with SelectionDAG.

arsenm commented on llvm/lib/CodeGen/CodeGenPrepare.cpp:
```diff
@@ -2298,8 +2298,24 @@ static bool despeculateCountZeros(IntrinsicInst *CountZeros,
+  //
+  // Despeculation is not only useless but also not wanted with SelectionDAG
+  // as XOR and CTLZ would be in different basic blocks.
+  if ((TLI->getTypeAction(CountZeros->getContext(), VTy) ==
+           TargetLowering::TypePromoteInteger ||
+       TLI->getOperationAction(ISD::CTLZ, VTy) == TargetLowering::Promote) &&
+      match(CountZeros->getOperand(0), m_Not(m_Value())))
+    return false;
```
Also move this below these other, simpler conditions?

v01dXYZ commented:
Thanks for your review.

Done:

* CGP and SD into two separate commits.
* only uses the `m/mi/sd_match` API. Prefer to use `getNOT/buildNoT` instead of manually creating `AllOnes` and `XOR` (this one is not available for VP though in SD).

The remaining task is the test for CGP (especially for vector types).

v01dXYZ commented:
I've got some problems when rebasing my work (Some unrelated tests are failing). That's why I won't fix today the conflict.

v01dXYZ commented:
I added some tests for CGP for two different cases:

* X86: DAG legalisation is the one doing the promotion
* RISCV: type legalisation is the one doing the promotion.

There is still a remaining issue with X86 when LZCNT is not available as DAG legalisation is custom for i8. I tried to mark `addOperationPromotedToType(CTLZ, i8, i32)` but the code is still not optimal with a spurious `XOR 31`.

Note also that there is no optimisation for counting trailing ones here, only for counting leading ones, although CGP takes the two cases into account in this PR. Tell me if you think it is relevant to add it.

v01dXYZ commented on llvm/lib/CodeGen/CodeGenPrepare.cpp:
```diff
@@ -2310,6 +2310,26 @@ static bool despeculateCountZeros(IntrinsicInst *CountZeros,
+  // Despeculation is not only useless but also not wanted with SelectionDAG
+  // as XOR and CTLZ/CTTZ would be in different basic blocks.
+  EVT VTy = TLI->getValueType(*DL, Ty);
+  int ISDOpcode = IntrinsicID == Intrinsic::ctlz ? ISD::CTLZ : ISD::CTTZ;
```
No because of the check above early exiting if `Operand[1] ~= One`.

topperc commented on llvm/lib/CodeGen/SelectionDAG/LegalizeIntegerTypes.cpp:
```diff
@@ -646,6 +648,39 @@ SDValue DAGTypeLegalizer::PromoteIntRes_Constant(SDNode *N) {
+// (CTLZ (XOR Op -1)) --> (CTLZ_ZERO_UNDEF (XOR (SHIFT (ANYEXTEND Op1)
+//                                                     ShiftAmount)
+//                                               -1))
+static bool ExtendCtlzNot(SDNode *Node, SDValue &Result, SDLoc &dl, EVT OVT,
```
Return an SDValue instead of using a reference parameter. You can return a default SDValue for the current `return false` case.

v01dXYZ commented:
Ping

v01dXYZ commented:
Ping

topperc commented on llvm/lib/CodeGen/CodeGenPrepare.cpp:
```diff
@@ -2310,6 +2310,26 @@ static bool despeculateCountZeros(IntrinsicInst *CountZeros,
+  // Despeculation is not only useless but also not wanted with SelectionDAG
+  // as XOR and CTLZ/CTTZ would be in different basic blocks.
+  EVT VTy = TLI->getValueType(*DL, Ty);
+  int ISDOpcode = IntrinsicID == Intrinsic::ctlz ? ISD::CTLZ : ISD::CTTZ;
```
Why `int` instead of `unsigned`?

v01dXYZ replied:
In this file, we have in other parts `int ISDOpcode = ...`, that's why I use `int`. But it seems there is no usage of setting a variable with an `ISD::<Variant>` with an `int` in the code base.

topperc commented on llvm/test/CodeGen/RISCV/rvv/ctlz-vp.ll:
```diff
@@ -2624,6 +2624,58 @@ define <vscale x 1 x i9> @vp_ctlz_zero_undef_nxv1i9(<vscale x 1 x i9> %va, <vsca
+; CHECK-ZVBB-NEXT:    vnot.v v8, v8, v0.t
+; CHECK-ZVBB-NEXT:    vclz.v v8, v8, v0.t
+; CHECK-ZVBB-NEXT:    ret
+  %va.not = xor <vscale x 1 x i9> %va, splat (i9 -1)
```
It would be very unusual to see a vp.ctlz and a regular xor. It be more likely to be a vp.xor and a vp.ctlz.

v01dXYZ replied:
I added support for `vp.xor` when it has the same mask and vector length as `vp.ctlz`.

v01dXYZ replied:
Maybe you want me to not try to match `VP_CTLZ + XOR` and only keep `VP_CTLZ + VP_XOR`.

topperc replied:
> Maybe you want me to not try to match VP_CTLZ + XOR and only keep VP_CTLZ + VP_XOR ?

Yeah not much point in matching a mixing of VP and non-VP. It won't scale well if we try to do that with every optimization.

Commit:
Count leading ones promoted type optimisation: Support VP_XOR

topperc commented on llvm/test/CodeGen/AArch64/ctlo.ll:
```diff
@@ -0,0 +1,100 @@
+  ret i8 %tmp2
+}
+
+define i8 @ctlo_i8_undef(i8 %x) {
```
undef -> poison. The flag is called `is_zero_poison`. It used to be `is_zero_undef`. Probably old tests didn't get renamed appropriately but we should use the right name for new tests.

topperc commented on llvm/test/CodeGen/AArch64/ctlo.ll:
```diff
@@ -0,0 +1,100 @@
+define i8 @ctlo_i8_undef(i8 %x) {
+; CHECK-LABEL: ctlo_i8_undef:
+; CHECK:       // %bb.0:
+; CHECK-NEXT:    mov w8, #-1 // =0xffffffff
```
This looks worse than the current codegen

```
ctlo_i8_undef:                          // @ctlo_i8_undef                        
        .cfi_startproc                                                           
// %bb.0:                                                                        
        mvn     w8, w0                                                           
        lsl     w8, w8, #24                                                      
        clz     w0, w8                                                           
        ret
```

topperc commented:
Have you investigated doing this as a DAGCombine after legalization? Its very unusual to change type legalization behavior based on surrounding instructions.

v01dXYZ commented:
@topperc Currently, the type legaliser not only extends the operands but needs to add a substract node too. Using a DAGCombiner would mean detect this pattern and reverse it instead of trying to prevent to create it (as it is in this PR). I understand your point though as this PR adds another exception to Type Legalisation. I'm not experienced enough to have a worthy opinion about which one is better.

I'll try to implement it with a DAGCombiner in another PR.

Commit:
Counting leading ones: Optimise only for CTLZ

Do not optimise for CTLZ_ZERO_UNDEF

Commit:
Counting leading ones: Optimise VP_CTLZ only with VP_XOR

Do not optimise for VP_CTLZ + XOR. That way, the non-VP and VP cases
are more clearly separated which is easier to follow.
